cilium/CFP-30788-managing-hubble-metrics-cardinality.md

```diff
@@ -0,0 +1,130 @@
+  - `contextoptions`: List of [context options](https://docs.cilium.io/en/latest/observability/metrics/#context-options) to be applied to the metric.
+  - `labels`: List of labels to be applied to the metric.
+  - `includeFilters` : List of include flow filters to be applied to this particular metric.
+  - `excludeFilters` : List of exclude flow filters to be applied to this particular metric.
+
+### Example
+    
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+    name: hubble-metrics-config
+    namespace: kube-system
+data:
+    hubblemetrics.yaml: |
+        hubbleMetrics:
+            enabled: true
+            metrics:
+            - name: flow
+                contextoptions: sourceEgressContext=pod;destinationIngressContext=pod
+                includeFilters: []
+                excludeFilters:
+                    destination_pod: ["kube-system/"]
+            - name: dns
+                contextoptions: labelsContext=source_namespace,source_pod
+                includeFilters: 
+                    source_pod: ["namespace1/"]
+```
+
+This will make sure we will have a dynamic way to manage each metric's cardinality, dimensions and targets. The flow filters can  be applied on each metric's `ProcessFlow` function save CPU cyles.
+
+
+## Impacts / Key Questions
+
+### Impact: Managing Hubble Metrics Cardinality
+
+This proposal intents to give more control of metrics to the cluster admins and developers. This will help in managing the cardinality of the metrics dynamically based on the cluster state without having to restart cilium agent or prometheus server. 
+
+### Impact: Dashboarding and Alerting
+
+### Will this proposal affect the existing dashboards and alerting rules?
+
+By keeping the existing behavior of reading cilium-config for `hubble-metrics` for backward compatibility, this proposal will not affect the existing dashboards and alerting rules.
+
+### Will this proposal affect dashboards built using the dynamic metrics config?
+
+Yes there is a possibility that the dashboards built using the dynamic metrics config may be affected if the properties of the metrics are changed dynamically. This will be a trade-off between the dynamic management of metrics and the stability of the dashboards. The cluster admins and developers should be aware of this trade-off.
+
+Also, this problem exists even with the existing behavior of reading cilium-config for `hubble-metrics`. New context options could be added to the metrics which may not be supported by the existing dashboards. 
+
+## Implementation Alternatives
+
+### Option 1: Move only the flow filters to the dynamic metrics config
+
+
+#### Pros
+
+* This will reduce the re-learning aspect of existing way to configure hubble metrics.
+
+#### Cons
+
+* It creates a split brain scenario where some of the same metric properties are configured in cilium-config and some in hubble-metrics-config.
+
+### Option 2: Introduce annotations to manage the cardinality of the metrics
```

Comment:
Not exactly the same, but I believe with some well-crafted `FlowFilters`, you could in theory do something similar with pod labels using your main proposal. Eg: you could write a config like:
```
- name: flow
  includeFilters:
    - source_label: ["hubble-flow-metrics=true"]
```

Which would then mean any pod with the label `hubble-flow-metrics` and a value of `true`  will produce flow metrics. 

That's one reason I like the config file approach; admins can basically centralize config; or craft configs that let users manage their metrics (or some combination of the two).

Reply:
That is a good idea, so this proposal infact is more customizable per admins' requirements on how they want to expose it.

Reply:
Exactly

Reply:
Can we also add a global source label to allow all metrics in a given pod. If that global source label is set as a Pod annotation then we observe all metrics

Reply:
Big :+1: for using annotations to configure metrics. Long term this is going to be the most flexible. There are a number of other configuration options that make sense at this level and would be easy to opt into with annotations 
- unidirectional flows - right now flow metrics are captured for both directions in and out of a workload. For highly connected workloads it would be useful to potentially capture only metrics in one direction. This would reduce the cardinality by half while still maintaining most the visibility.
- verdict filtering - again for high cardinality workloads, it might be useful only to track metrics for given verdicts.  omitting forwarded verdicts and only shipping metrics for dropped verdicts that might be useful to have alerts on.